maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java

```diff
@@ -1034,12 +1035,19 @@ private int execute( CliRequest cliRequest )
             List<MavenProject> sortedProjects = result.getTopologicallySortedProjects();
             if ( result.canResume() )
             {
-                logBuildResumeHint( "mvn <args> -r " );
+                logBuildResumeHint( "mvn <args> -r" );
             }
-            else if ( project != null && !project.equals( sortedProjects.get( 0 ) ) )
+            else if ( !failedProjects.isEmpty() )
             {
-                String resumeFromSelector = getResumeFromSelector( sortedProjects, project );
-                logBuildResumeHint( "mvn <args> -rf " + resumeFromSelector );
+                // Sort the failedProjects list in the topologically sorted order.
+                failedProjects.sort( comparing( sortedProjects::indexOf ) );
+
+                MavenProject firstFailedProject = failedProjects.get( 0 );
+                if ( !firstFailedProject.equals( sortedProjects.get( 0 ) ) )
```

Comment:
Either we're missing an `else` branch here, or the condition will always evaluate to true due to the `canResume()` check before?

Reply:
We can only hit this line when `result.canResume()` returned `false` (if it returned `true`, we would've hit line 1038). If the first failed project is not the first in topological order, the build would have been resumable, and thus we'd never get here. @MartinKanters, I think this check can be removed, would you agree?

Reply:
I think there is a rare situation where this check would be valid:
resume.properties could not be persisted somehow ( `setCanResume( true )` will not be called) and then we still need this check to verify that the first project succeeded (otherwise -rf does not make sense).

It's a very much edge case, though. We could remove it altogether, but let's first monitor for some time if -r is performing well?

Reply:
Else branch is not needed: in that case, the user would just run the same command as they did the last time, so no need to hint them about resuming the build.